### ISSUE TITLE
Configuring the DataProtection options via appsettings.json file

### ISSUE THREAD

**Issue by joymon** [dotnet/aspnetcore#45942]
## Summary

Currently, it seems the DataProtection settings need to be configured via code. We are looking for support via appsettings.json or in general via configuration without recompiling the application for different deployment environment types.

## Motivation and goals

As developers, we would like to code once and deploy in various environment types by tuning the configuration. Currently, the DataProtection configuration seems to support only via code. Though we can read configuration in a custom way, it would be great if the library provides the configuration options out of box.

## In scope

Configure the DataProtection options from appsettings.json. In code, we will be adding `services.AddDataProtection();`

## Risks / unknowns

How might developers misinterpret/misuse this? How might implementing it restrict us from other enhancements in the future? Also list any perf/security/correctness concerns.

## Examples

`services.AddDataProtection()` - This will be loading the settings from the appsettings.json before defaulting to the values in the code.

ex: By default, the keys are stores at %USERPROFILE%\AppData\Local\ASP.NET\DataProtection-Keys. To change this folder we have to write code. This path should be configurable via appsettings.json

**Comment by davidfowl** [comment]
Can you provide more details? Are you just trying to change this one thing? Can you show what the code looks like today and what you'd like the setting to express?

**Comment by joymon** [comment]
The current way we are enabling the DataProtection is `services.AddDataProtection()` . When it runs on a desktop, it creates the key file %USERPROFILE%\AppData\Local\ASP.NET\DataProtection-Keys. When it runs in a load-balanced environment, it still creates files in the same place and fails as other machines has different keys. In load balanced environment, we have to point to a \\SharedSMB path. It is possible via `services.AddDataProtection().PersistKeysToFileSystem(new DirectoryInfo(@"\\SMBSharedPath"));` but it needs code change. We are trying to use only `services.AddDataProtection()` and rest to be configurable from appsettings.json. Hope it clears

**Comment by davidfowl** [comment]
Sure, but you can write this code:

```C#
var path = configuration["DataProtection:Path"];
services.AddDataProtection().PersistKeysToFileSystem(path);
```

**Comment by joymon** [comment]
Hi @davidfowl,
As mentioned in the original question, we can do the custom way by reading the configuration. But looking for an official / out-of-the-box way to read the config.  Similar to the "Logging" section in appsettings.json. Does the DataProtection supports reading the configuration out of the box as of now?

**Comment by joymon** [comment]
Response inline
- What options can you configure _[Joy] We are looking for configuring the location of keys. But if you are designing I would recommend all the settings that can be done via code should be present in configuration_
- How do you get this behavior by default? Is it opt-in or out.  _[Joy] Default behavior is to read from config similar to logging_
- Is there a new API?_[Joy] Not sure what you mean. Not expecting any new .Net class or method. The `services.AddDataProtection()` is expected to read from config_

**Comment by blowdart** [comment]
The problem with using config is there is no common config between the providers, unlike the logging example @davidfowl talks about. The file path is unique to the file keyring provider, the key vault location is unique to keyvault etc.

Given there's a way to check environments already that would be the approach I'd take in my program.cs

**Comment by davidfowl** [comment]
I think it's reasonable to provide a way to configure this (answering some of [these questions](https://github.com/dotnet/aspnetcore/issues/45942#issuecomment-1375971967))

**Comment by joymon** [comment]
> The problem with using config is there is no common config between the providers, unlike the logging example @davidfowl talks about. The file path is unique to the file keyring provider, the key vault location is unique to keyvault etc.
> 
> Given there's a way to check environments already that would be the approach I'd take in my program.cs

@blowdart I agree that the settings are different for different providers. Though not exactly the same, the logging also works similarly. If we are logging into the console, no or very fewer options, but if it's to Azure App Insights, it requires an instrumentation key. I am not able to envision how the JSON should look as I am not aware of all the options. Once I am free from the busy schedule I will put some through. Hope by the time the current team get some time to look at this featue.
